react/src/hooks/useControllableState.ts

```diff
@@ -0,0 +1,135 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import type { SetStateAction } from 'react';
+
+export interface Options<T> {
+  defaultValue?: T;
+  defaultValuePropName?: string;
+  valuePropName?: string;
+  trigger?: string;
+}
+
+export type Props = Record<string, any>;
+
+export interface StandardProps<T> {
+  value: T;
+  defaultValue?: T;
+  onChange: (val: T) => void;
+}
+/**
+ * Custom hook to  manage this kind of state. managed by itself or controlled by it's parent
+ * useControllableState is based on useControllableValue.
+ * However if the value is undefined, the component is treated as an uncontrolled component
+ */
+// params 설명 더 작성하기
+
+function useControllableState<T = any>(
+  props: StandardProps<T>,
+): [T, (v: SetStateAction<T>) => void];
+function useControllableState<T = any>(
+  props?: Props,
+  options?: Options<T>,
+): [T, (v: SetStateAction<T>, ...args: any[]) => void];
+function useControllableState<T = any>(
+  props: Props = {},
+  options: Options<T> = {},
+) {
+  const {
+    defaultValue,
+    defaultValuePropName = 'defaultValue',
+    valuePropName = 'value',
+    trigger = 'onChange',
+  } = options;
+
+  const value = props[valuePropName] as T;
+
+  let isControlledRef = useRef(value !== undefined);
+  let isControlled = value !== undefined;
+
+  useEffect(() => {
+    let wasControlled = isControlledRef.current;
+    if (wasControlled !== isControlled) {
+      console.warn(
+        `WARN: A component changed from ${wasControlled ? 'controlled' : 'uncontrolled'} to ${
+          isControlled ? 'controlled' : 'uncontrolled'
+        }.`,
+      );
+    }
+    isControlledRef.current = isControlled;
+  }, [isControlled]);
+
+  const initialValue = useMemo(() => {
+    if (isControlled) {
+      return value;
+    }
+    if (Object.prototype.hasOwnProperty.call(props, defaultValuePropName)) {
+      return props[defaultValuePropName];
+    }
+    return defaultValue;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const stateRef = useRef(initialValue);
+  if (isControlled) {
+    stateRef.current = value;
+  }
+
+  const update = useUpdate();
+
+  function setState(v: SetStateAction<T>, ...args: any[]) {
+    const r = isFunction(v) ? v(stateRef.current) : v;
+
+    if (!isControlled) {
+      stateRef.current = r;
+      update();
+    }
+    if (props[trigger]) {
+      props[trigger](r, ...args);
+    }
+  }
+
+  return [stateRef.current, useMemoizedFn(setState)] as const;
+}
+
+export default useControllableState;
+
+function isFunction(value: unknown): value is (...args: any) => any {
+  return typeof value === 'function';
+}
```

Comment:
you can use lodash's [isFunction](https://lodash.com/docs/4.17.15#isFunction)

Reply:
I modified it https://github.com/lablup/backend.ai-webui/pull/2572/files#diff-e98538d8ef63e2ebd590961cc47a0a8badea1d72e19f7cf0cbe5a3501e0a97bbR79